### PR TITLE
[WIP] Depth-Aware Domain Adaptation (DADA)

### PR DESCRIPTION
**Pull request by alexrey88**
Use approach from [DADA paper](https://github.com/valeoai/DADA) to improve segmentation head, and thereby the masker. 

Summary of changes:

- Feature fusion: features obtained from depth decoder, just before average pooling, are now passed through a small decoder to obtain an output of the same size as latent vector z. We then compute the element-wise product of latent vector z and depth_features. The resulting vector is given as input to the segmentation decoder.
- DADA fusion: the weighted self-information map of segmentation output is now multiplied by the depth output before being passed through the ADVENT discriminator of segmentation head. (Idea is that now, closer objects are given more attention).

### CODE REVIEW

File: omnigan/generator.py [hunk 1]
```diff
@@ -132,20 +130,54 @@ def __init__(self, opts):
...
+        depth = F.interpolate(
+            depth,
+            size=(384, 384),  # size used in MiDaS inference
+            mode="bicubic",  # what MiDaS uses
+            align_corners=False,
+        )
+        depth = F.interpolate(
+            depth, (self.output_size, self.output_size), mode="nearest"
+        )  # what we used in the transforms to resize input
+
```
Comment (alexrey88) [inline omnigan/generator.py:171]: Only lines added from previous DepthDecoder are lines 172-175

File: omnigan/trainer.py [hunk 2]
```diff
@@ -214,9 +214,7 @@ def resume_from_path(
         if new_exp:
             exp = Experiment(project_name="omnigan", **comet_kwargs)
             exp.log_asset_folder(
-                str(Path(__file__).parent),
-                recursive=True,
-                log_file_name=True,
+                str(Path(__file__).parent), recursive=True, log_file_name=True,
             )
```
Comment (alexrey88) [inline omnigan/trainer.py:218]: Don't know why it's doing this again..

File: omnigan/generator.py [hunk 3]
```diff
@@ -136,20 +134,54 @@ def __init__(self, opts):
...
+class DepthDecoder(nn.Module):
+    """ Depth decoder based on depth auxiliary task in DADA paper
+        Source: https://github.com/valeoai/DADA
+    """
+
+    def __init__(self, opts):
+        super().__init__()
+        self.relu = nn.ReLU(inplace=True)
+        self.enc4_1 = nn.Conv2d(
+            2048, 512, kernel_size=1, stride=1, padding=0, bias=True
```
Comment (vict0rsch) [inline omnigan/generator.py:146]: this should be `opts.gen.d.res_dim` as it will change with MOAPR

Comment (vict0rsch) [reply omnigan/generator.py:146]: Actually if you look at MOAPR there's also a `mid_dim` var

File: shared/trainer/defaults.yaml [hunk 4]
```diff
@@ -108,6 +108,8 @@ gen:
     output_dim: 1
     conv_norm: batch
     res_norm: batch
+    loss_name: sigm # choice between "sigm" or "berhu" (for reverse Huber loss)
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:111]: 😄

File: shared/trainer/defaults.yaml [hunk 5]
```diff
@@ -136,6 +138,7 @@ gen:
     use_minent_var: True # add variance of entropy map in the measure of entropy for a certain picture
     use_advent: True # minimize the entropy of the image by adversarial training
     use_ground_intersection: True
+    do_feat_fusion: false
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:141]: why would you have use_dada and not do_feat_fusion?

**Review by vict0rsch — CHANGES_REQUESTED** [review]

File: shared/trainer/defaults.yaml (hunk 5 above)
Comment (alexrey88) [reply shared/trainer/defaults.yaml:141]: hmmm i've put `use_dada `in gen.d to tell that we're using DADA as presented in the paper (so feat fusion + dada fusion on seg head), and I've put `do_feature_fusion `for the masker instead of use_dada to tell that we're just doing feature fusion (and not dada fusion also). You think I should maybe have `do_feature_fusion `and `do_dada_fusion `in both gen.s and gen.m instead? Wasn't sure how to organize that.

Comment (vict0rsch) [reply shared/trainer/defaults.yaml:141]: oh ok I see dada is for the seg head and do_feat_fusion is for the mask head
